practical/ProcessMining/group2/alphaplusminer/alpha_plus_miner.py

```diff
@@ -0,0 +1,479 @@
+                            and relation_between_pair != relation):
+                            makePair = False
+                            break
+                        else:
+                            relation_between_pair = relation
+                        if relation != "->":
+                            makePair = False
+                            break
+                if makePair is True:
+                    if relation_between_pair == "->":
+                        new_pair = (pair_choices1, pair_choices2)
+                    else:
+                        new_pair = (pair_choices2, pair_choices1) # TODO- include in tests
+                    pairs.append(new_pair)
+
+        self.pairs = pairs
+
+
+
+    def get_maximal_pairs(self):
+            # Set of paired activities that are maximal
+            pos1 = 0
+            pair_appended = []
+            maximal_pairs = []
+            for pair1 in self.pairs:
+                append_flag = True
+                # flatten pair 1
+                flat_pair1 = []
+                for pair_element in pair1:
+                    for e in pair_element:
+                        flat_pair1.append(e)
+
+                pos2 = 0
+                for pair2 in self.pairs:
+                    if pos1 != pos2:
+                        flat_pair2 = []
+                        for pair_element in pair2:
+                            for e in pair_element:
+                                flat_pair2.append(e)
+                        # flat the pair 1
+
+                        # flat the pair 2
+                        # check if pair1 issubset of pair 2 or pair 2 is subset of 1
+                        if (SortedSet(flat_pair1).issubset(flat_pair2) 
+                        and SortedSet(flat_pair1) != SortedSet(flat_pair2)):
+                            append_flag = False
+                    pos2 = pos2 + 1
+
+                if append_flag is True:
+
+                    if SortedSet(flat_pair1) not in pair_appended:
+                        maximal_pairs.append(pair1)
+                        pair_appended.append(SortedSet(flat_pair1))
+                pos1 = pos1 + 1
+
+            self.maximal_pairs = maximal_pairs
+
+
+
+
+
+
+
+    def get_length_one_loops(self) -> SortedSet:
+        # extract length one loop
+        self.length_one_loops = SortedSet()
+        self.get_transitions()
+        # compute footprint matrix and extract all transitions that have a causality relation with itself, eg: aa, bb etc.
+
+        self.get_footprint()
+
+        for transition in self.transitions:
+            if self.relations[transition][transition] == "->":
+                self.length_one_loops.add(transition)
+
+
+
+    def remove_length_one_loops(self) -> SortedSet:
+        # This function corresponds to step 3 of the alpha plus algorithm:  3) T' := T \ L1L
+        self.log_without_length_one_loops = self.transitions.difference(self.length_one_loops)
+
+
+
+
+
+    def get_FL1L(self):
```

Comment:
try to use a descriptive method title instead of copying from the paper. Without the background knowledge I had no idea what to expect from this